sender.go

```diff
@@ -184,6 +158,46 @@ func (s *sender) trySend(ctx context.Context, evt eventer) error {
 	}
 }
 
+func (s *sender) recover(ctx context.Context, err error) error {
+	sp, ctx := s.startProducerSpanFromContext(ctx, "eh.sender.recover")
+	defer sp.Finish()
+
+	select {
+	case <-ctx.Done():
+		// context is done, so return
+		log.For(ctx).Error(ctx.Err())
+		return ctx.Err()
+	default:
+		err := s.Recover(ctx)
+		if err == nil {
+			// recovered
+			log.For(ctx).Debug("recovered connection")
+			return nil
+		}
+
+		if isRecoverableError(err) {
+			log.For(ctx).Debug("retrying error: " + err.Error())
+			return common.Retryable(err.Error())
+		}
+
+		// failed recovery... something is broken
+		log.For(ctx).Error(err)
+		return err
+	}
+}
+
+func isRecoverableError(err error) bool {
+	if _, ok := err.(*amqp.Error); ok {
+		return true
+	} else if _, ok = err.(*amqp.DetachError); ok {
+		return true
+	} else if _, ok = err.(amqp.DetachError); ok {
+		return true
+	} else if _, ok = err.(amqp.Error); ok {
+		return true
+	}
+	return false
+}
```

Comment:
@devigned I assume these are originating from pack.ag/amqp? I consider inconsistent errors types a bug and opened https://github.com/vcabbage/amqp/issues/113. I'll dig through to see what's causing them later, but if you have any details readily available it would be appreciated.

Reply:
Wow, that was quick 👍 . I guess I don't need to open an issue at this point.

Let me tell you, this was a tough one to track down. It would occur when the SAS token used to negotiate the connection expired and subsequently the link was detached.

```
time="2018-07-24 23:06:36" level=error msg="link detached, reason: *Error(nil)
```